File: FirstProjectAMIB1/states.cpp
#include "states.h"

static const StateInfo state_infos[3] = {
  {IDLE::setup, IDLE::enter, IDLE::exit, IDLE::loop, IDLE::event},
  {MOTIONMACHINE::setup, MOTIONMACHINE::enter, MOTIONMACHINE::exit, MOTIONMACHINE::loop, MOTIONMACHINE::event},
  {ARM::setup, ARM::enter, ARM::exit, ARM::loop, ARM::event}
};

static const WireValue wire_values[2] = {
  {1, 0, sizeof(uint32_t), (Value<void*>*) &MOTIONMACHINE::stepperPosition},
  {2, 0, sizeof(uint32_t), (Value<void*>*) &ARM::rotations}
};

MasterManager<State, 3, 2> manager(0x19465309, state_infos, wire_values, 0);

namespace IDLE {


void event(uint8_t ev) {
  switch (ev) {
  
  default:
    break;
  }
}


}
namespace MOTIONMACHINE {
Value<uint32_t> stepperPosition;

void event(uint8_t ev) {
  switch (ev) {
  case 0:
    events::moveLiftUp();
    break;
  case 1:
    events::moveToBottom();
    break;
  case 2:
    events::setLiftToZero();
    break;
  case 3:
    events::runSteps();
    break;
  case 4:
    events::stopSteps();
    break;
  default:
    break;
  }
}


}
namespace ARM {
Value<uint32_t> rotations;

void event(uint8_t ev) {
  switch (ev) {
  case 0:
    events::moveFromTallToShort();
    break;
  case 1:
    events::moveFromShortToTall();
    break;
  case 2:
    events::disableElectromagnet();
    break;
  case 3:
    events::enableElectromagnet();
    break;
  case 4:
    events::lowerArm();
    break;
  case 5:
    events::raiseArm();
    break;
  case 6:
    events::resetArmPosition();
    break;
  case 7:
    events::moveArm();
    break;
  default:
    break;
  }
}

namespace tablet {

namespace events {
void finishedAction() { manager.sendTabletEvent(0); }
}
}
}

File: states.js

var IDLE = {
  id: 0,
  master: {
    values: {
      
    },
    events: {
      
    }
  },
  tablet: {
    values: {
      
    },
    events: {
      
    }
  }
};
var STATE_IDLE = 0;
var MOTIONMACHINE = {
  id: 1,
  master: {
    values: {
      stepperPosition: new HardwareValue(1, 0, Manager.TYPE_UINT32)
    },
    events: {
      moveLiftUp: function moveLiftUp() { manager.sendEvent(0, 1); },
      moveToBottom: function moveToBottom() { manager.sendEvent(1, 1); },
      setLiftToZero: function setLiftToZero() { manager.sendEvent(2, 1); },
      runSteps: function runSteps() { manager.sendEvent(3, 1); },
      stopSteps: function stopSteps() { manager.sendEvent(4, 1); }
    }
  },
  tablet: {
    values: {
      
    },
    events: {
      finishedAction: new LocalEvent(1, 0)
    }
  }
};
var STATE_MOTIONMACHINE = 1;
var ARM = {
  id: 2,
  master: {
    values: {
      rotations: new HardwareValue(2, 0, Manager.TYPE_UINT32)
    },
    events: {
      moveFromTallToShort: function moveFromTallToShort() { manager.sendEvent(0, 2); },
      moveFromShortToTall: function moveFromShortToTall() { manager.sendEvent(1, 2); },
      disableElectromagnet: function disableElectromagnet() { manager.sendEvent(2, 2); },
      enableElectromagnet: function enableElectromagnet() { manager.sendEvent(3, 2); },
      lowerArm: function lowerArm() { manager.sendEvent(4, 2); },
      raiseArm: function raiseArm() { manager.sendEvent(5, 2); },
      resetArmPosition: function resetArmPosition() { manager.sendEvent(6, 2); },
      moveArm: function moveArm() { manager.sendEvent(7, 2); }
    }
  },
  tablet: {
    values: {
      
    },
    events: {
      finishedAction: new LocalEvent(2, 0)
    }
  }
};
var STATE_ARM = 2;

var STATES = {
  IDLE: IDLE,
  MOTIONMACHINE: MOTIONMACHINE,
  ARM: ARM
};
var manager = new Manager([IDLE, MOTIONMACHINE, ARM]);
